articles/ros_node_idl.md

```diff
@@ -0,0 +1,187 @@
+---
+layout: default
+title: ROS 2 Node Interface Definition Language
+permalink: articles/ros2_node_interface_definition_language.html
+abstract:
+  This article specifies the ROS 2 Interface Definition Language, a simple and standardized manner to export the complete interface (action/message/parameter/service) of node(s) in a package.
+author: >
+  [Jérémie Deray](https://github.com/artivis),
+  [Kyle Fazzari](https://github.com/kyrofa)
+published: true
+categories: Interfaces
+---
+
+{:toc}
+
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+
+## Background
+
+Every ROS 2 node has an associated interface that describes how it communicates with other nodes, as well as how it is to be configured.
+This interface is defined in code, and consists of:
+
+- Actions (server or client)
+- Parameters
+- Services (server or client)
+- Topics (publisher or subscriber)
+
+The information contained within that interface is obviously very valuable on different levels and from different perspectives.
+While it is usually readily available to a developer looking at the code, it cannot reliably be automatically extracted.
+It therefore calls for the creation of a standardized way to explicitly define and export this information.
+
+This article defines a high-level abstraction allowing upstream packages to specify the communication requirements of the nodes in the package, such that the final user, be it a developer or a static analysis tool, can benefit from it.
+The Interface Definition Language (IDL) specified in the next section is meant to be distributed alongside its associated package, be it in the source code or a generated release packaging format (e.g. debian).
```

Comment:
So I see the potential for confusion, but what is proposed does seem like it could be called an IDL. For nodes. So *Node IDL* seems appropriate.

Reply:
I believe the terminology is appropriate, however I do agree that it could be confusing. As discussed in another thread with @dirk-thomas we leave the potential renaming to later on.

ps - I like 'NoDL', other acronymes include 'NAPI', 'NAPID' (Node API Description).

Reply:
I didn't mean to imply the terminology is incorrect, it is perfectly fine. But it's an over-used term and a catchy name avoids that. :)

Reply:
Yeah, no strong feelings on the name. It is indeed an IDL, but we can name it whatever we like.